Fastsurfer Singularity workflow to quick test new code before merging

taha-abdullah opened the pull request:
This PR adds a workflow to test new code inside a Singularity container, running on a local runner, before it is merged into the `dev` branch. It also adds some quick tests to evaluate the new code and includes documentation for how to set up and run the workflow and tests.

### **.github/workflows/**

**fastsurfer_singularity.yaml**
- workflow to build and test new code in a Singularity container running on a local GitHub Actions runner

**singularity_workflow.md**
- dcumentation explaining how to run fastsurfer_singularity.yaml

### **test/**

**quick_test/**
- **test_file_existence.py**
  - unittest to check if all necessary files are present in FastSurfer output

- **test_errors.py**
  - unittest to check if errors are present in FastSurfer log files

**data/**
- **files.yaml**
  - YAML file containing filenames that should be present in FastSurfer output

- **errors.yaml**
  - YAML file containing error criteria

Switching from Docker to Singularity:
The workflow originally used Docker but was switched to Singularity later on.

taha-abdullah committed:
Fastsurfer Singularity workflow to quick test new code before merging

Removed extra workflows used for local testing

Removed extra files used for local testing

documentation for the fastsurfer singularity test workflow

Delete test/__pycache__ directory

Update fastsurfer_singularity.yaml

converting tests to run on singularity

Switching from Docker to Singularity

changed runner to ci-gpu

Deleted old action files

Changed the name of FastSurfer actions to differentiate between Docker and Singularity

Resolved issues from code review

added fastsurfer_singularity.yaml to run FastSurfer in singularity container

fixed issues in test_file_existence

completed implementation of tests through command line arguments instead of yaml files

Taking yaml files using command line arguments

Implemented script to check for errors in log files

Adding a file existence test to make sure all the
files that are supposed to be in the results are
actually there.

making jobs sequential by adding dependencies

Added environment variables for MRI data and output

renaming the workflow to 'FastSurfer Docker Image Build'

Building GPU FastSurfer Image and running analysis

simple workflow to test if self hosted runner 'dirac' is running correctly

dkuegler commented on .github/workflows/fastsurfer_singularity.yaml:
```diff
@@ -0,0 +1,90 @@
+    runs-on: ci-gpu
+    needs: prepare-job
+    steps:
+    - name: Build Docker Image and conver to Singularity
```
```suggestion
    - name: Build Docker Image and convert to Singularity
```

m-reuter replied:
```suggestion
    - name: Build Docker Image and convert to Singularity
```

dkuegler commented on .github/workflows/fastsurfer_singularity.yaml:
I would probably call this file "quicktest.yaml".

dkuegler commented on .github/workflows/singularity_workflow.md:
name "QUICKTEST.md"

dkuegler commented on test/quick_test/test_errors.py:
```diff
@@ -0,0 +1,89 @@
+        for log_file in self.log_files:
+            lines_with_keywords = []
+            try:
+                with open(log_file, 'r') as file:
```
I think you can also do `file.readLines()` to get all lines. here

dkuegler commented on test/quick_test/test_errors.py:
```diff
@@ -0,0 +1,89 @@
+                    for line_number, line in enumerate(file, start=1):
+                        if any(keyword in line.lower() for keyword in self.errors):
+                            if not any(white in line.lower() for white in self.whitelist):
+                                lines_with_keywords.append((line_number, line.strip()))
```
And here, it would make sense to add 2 lines before and 2 lines after for context.

dkuegler commented on test/quick_test/test_errors.py:
```diff
@@ -0,0 +1,89 @@
+
+        try:
+            # Retrieve the log files in given log directory
+            cls.log_files = [os.path.join(cls.log_directory, file)
```
I prefer using `pathlib.Path` over `os.path`.

dkuegler commented on test/quick_test/test_errors.py:
```diff
@@ -0,0 +1,89 @@
+    """
+
+    parser = argparse.ArgumentParser(description="Test for errors in log files.")
+    parser.add_argument('log_directory', type=str, help="The directory containing the log files.")
```
```suggestion
    parser.add_argument('log_directory', type=Path, help="The directory containing the log files.")
```

also needs `from pathlib import Path` and some changes above.

dkuegler commented on test/quick_test/test_errors.py:
```diff
@@ -0,0 +1,89 @@
+
+    TestErrors.log_directory = args.log_directory
+
+    unittest.main(argv=[sys.argv[0]])
```
make sure all files end on an empty line.

taha-abdullah committed:
errors.yaml:
- added errors and whitelists

QUICKTEST.md:
- renamed from singularity_workflow.md

quicktest.yaml:
- renamed from singularity_workflow.md
- fixed typo

test_errors.py:
- using pathlib.Path instead of os.path
- added context for errors
- improved error output to give more information about the errors

test_file_existence.py:
- using pathlib.Path instead of os.path

in test-error-messages:
- changed file name to test/quick_test/test_errors.py

dkuegler commented on test/quick_test/test_file_existence.py:
some comments should be applied here same as other unit test. (e.g. Path)

dkuegler commented on test/quick_test/test_errors.py:
```diff
@@ -0,0 +1,89 @@
+    A test case class to check for the word "error" in the given log files.
+    """
+
+    error_file_path = "./test/quick_test/data/errors.yaml"
```
```suggestion
    error_file_path: Path = Path("./test/quick_test/data/errors.yaml")
```

m-reuter commented on .github/workflows/singularity_workflow.md:
```diff
@@ -0,0 +1,57 @@
+
+### Build Singularity Image
+
+This job builds a Docker image and converts it to a Singularity image. It depends on the successful completion of the `prepare-job`. The Docker image is built using a Python script `Docker/build.py` with the `--device cuda --dry_run --tag fastsurfer_gpu:cuda` flags. The Docker image is then converted to a Singularity image.
```
--dry_run will not actually build the docker image?

dkuegler replied:
`--dry_run` will not build the image, but `build.py [...] --dry_run | bash` will.

It is probably better to not use the `--dry_run` option here, which does all the steps in the background plus prints the command it would run to stdout.

dkuegler replied:
That is also, why technically the plag `--print` is a better alias to `--dry_run`.

taha-abdullah replied:
--dry_run by itself only outputs the docker build command but does not execute it. I am piping the command to bash in order to build the docker image

m-reuter replied:
so why not use command without dry_run directly then?

taha-abdullah committed:
update conf.py

- replacing hardcoded 'stable' branch with the current working branch so docs point to current branch code

taha-abdullah committed:
Revert "update conf.py"

This reverts commit 813be1203215dd3e25b8fc0f1503af15388326ff.

dkuegler commented on .github/workflows/QUICKTEST.md:
```diff
@@ -0,0 +1,57 @@
+
+### Build Singularity Image
+
+This job builds a Docker image and converts it to a Singularity image. It depends on the successful completion of the `prepare-job`. The Docker image is built using a Python script `Docker/build.py` with the `--device cuda --dry_run --tag fastsurfer_gpu:cuda` flags. The Docker image is then converted to a Singularity image.
```
```suggestion
This job builds a Docker image and converts it to a Singularity image. It depends on the successful completion of the `prepare-job`. The Docker image is built using a Python script `Docker/build.py` with the `--device cuda --tag fastsurfer_gpu:cuda` flags. The Docker image is then converted to a Singularity image.
```

dkuegler commented on .github/workflows/QUICKTEST.md:
```diff
@@ -0,0 +1,57 @@
+To use this workflow, you need to have a self-hosted runner labeled `ci-gpu` set up on your machine. You also need to update the environment variables of the runner, by going to `/home/your_runner/.env` file and adding the following environment variables with the actual paths you want to use.
+
+
+RUNNER_FS_MRI_DATA: Path to MRI Data
+
+RUNNER_FS_OUTPUT: Path to Output directory
+
+RUNNER_FS_LICENSE: Path to License directory
+
+RUNNER_SINGULARITY_IMGS: Path to where Singularity images should be stored
+
+RUNNER_FS_OUTPUT_FILES: Path to output files to be tested
+
+RUNNER_FS_OUTPUT_LOGS: Path to output log files to check for errors
```
```suggestion
### Environment variables
`RUNNER_FS_MRI_DATA`: Path to MRI Data

`RUNNER_FS_OUTPUT`: Path to Output directory

`RUNNER_FS_LICENSE`: Path to License directory

`RUNNER_SINGULARITY_IMGS`: Path to where Singularity images should be stored

`RUNNER_FS_OUTPUT_FILES`: Path to output files to be tested

`RUNNER_FS_OUTPUT_LOGS`: Path to output log files to check for errors
```

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -0,0 +1,90 @@
+          echo "SIF File does not exist. Building file."
+          PYTHONPATH=$PYTHONPATH
+          cd $PYTHONPATH
+          python3 Docker/build.py --device cuda --dry_run --tag fastsurfer_gpu:cuda | bash
```
```suggestion
          python3 Docker/build.py --device cuda --tag fastsurfer_gpu:cuda
```

dkuegler commented on .gitignore:
Files should end on a newline

dkuegler commented on test/quick_test/test_errors.py:
```diff
@@ -0,0 +1,99 @@
+                        if any(error in line.lower() for error in self.errors):
+                            if not any(white in line.lower() for white in self.whitelist):
+                                # Get two lines before and after the current line
+                                context = []
+                                for i in range(-3, 2):
+                                    if 0 <= line_number + i < len(lines):
+                                        context.append(lines[line_number + i])
```
```suggestion
                                context = lines[max(0, line_number-2):min(len(lines), line_number+3)]
```

dkuegler commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
-/.idea/**
\ No newline at end of file
+/.idea/**
+/rough_work/**
```
```suggestion
/rough_work/**

```

dkuegler commented on test/__init__.py:
```diff
@@ -0,0 +1,8 @@
+    'test_file_existence',
+    'test_error_messages',
+    'test_errors'
+    ]  
```
```suggestion
    ]  

```

dkuegler reviewed (approved):
Looks

dkuegler committed:
Apply suggestions from code review
